src/main/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiff.scala

```diff
@@ -73,21 +73,25 @@ object BasicOpcDiff extends OpcDiffer {
 
   def convertion(x: DataEntry, t: DataEntry): Either[ValidationError, DataEntry] =
     t.dataType match {
-      case DataTypeObj => x.dataType match {
-        case Int32 | Amount | Timestamp | BigInteger => {
+      case DataTypeObj =>(x.dataType match {
+        case Int32 | Amount | Timestamp => Right(x.data) 
+        case BigInteger => Right(x.data.drop(2))
+        case _ => Left(ContractUnsupportedOPC)
+      }) flatMap { bytes => {
           val to: DataTypeVal[_] = DataTypeObj.deserializer(t.data)
           to match {
-            case Int32      => formatResB[Int]   (x.data, Int32)
-            case Amount     => formatResB[Long]  (x.data, Amount)
-            case Timestamp  => formatResB[Long]  (x.data, Timestamp)
-            case BigInteger => formatResB[BigInt](x.data, BigInteger)
+            case Int32      => formatResB[Int]   (addLeadingZeros(bytes, 4), Int32)
+            case Amount     => formatResB[Long]  (addLeadingZeros(bytes, 8), Amount)
+            case Timestamp  => formatResB[Long]  (addLeadingZeros(bytes, 8), Timestamp)
+            case BigInteger => formatResB[BigInt](bytes.dropWhile(i => i == 0), BigInteger)
```

Comment:
It still has some problems, e.g.
`Ints.toByteArray(-2147483648)` => `Array(-128, 0, 0, 0)`
`Longs.toByteArray(-2147483648L)` => `Array(-1, -1, -1, -1, -128, 0, 0, 0)`

Reply:
Fixed by new commit

Reply:
Let's separate the cases: (Amount and Timestamp are all Long type, so just combined the cases to Amount)
1. Int32 and Amount
- Int32 => Amount
`-2147483648  => Ints.toByteArray(-2147483648) => Array(-128, 0, 0, 0) => Array(0, 0, 0, 0, -128, 0, 0, 0) => Longs.fromByteArray(Array(0, 0, 0, 0, -128, 0, 0, 0)) => 2147483648`
since negative value will add leading `-1` to the byte array(not sure, need to confirm)
- Amount => Int32
`-1 => Longs.toByteArray(-1) => Array(-1, -1, -1, -1, -128, 0, 0, 0) => ValidationError.OverflowError`
2. Int32/Amount and BigInt
`Ints.toByteArray(-1) = Array(-1, -1, -1, -1)`
`Longs.toByteArray(-1) = Array(-1, -1, -1, -1, -1, -1, -1, -1)`
`BigInt(-1).toByteArray = Array(-1)`
`Ints.toByteArray(-2147483648) = Array(-128, 0, 0, 0)`
`Longs.toByteArray(-2147483648L) = Array(-1, -1, -1, -1, -128, 0, 0, 0)`
`BigInt(-2147483648L).toByteArray = Array(-128, 0, 0, 0)`
And remove leading `-1` is also not the correct solution
`Longs.toByteArray(-2147483648L-1L) = Array(-1, -1, -1, -1, 127, -1, -1, -1)`
` BigInt(-2147483648L-1L).toByteArray = Array(-1, 127, -1, -1, -1)`
`Ints.fromByteArray(Array(127, -1, -1, -1)) = 2147483647`

Reply:
1. Add leading `-1` when first byte is negative is a effective way, I'll fix it in new commit.
2. `-2147483648L-1L` is out of range of Int, so `Array(127, -1, -1, -1)` is acceptable instead of an `overflow error`.
3.  `Array(-1, -1, -1, -1, 127, -1, -1, -1)` is also a correct byte array for `BigInt`, since `BigInt(Array[Byte](-1, -1, -1, -1, 127, -1, -1, -1))` is just the same as `BigInt(-2147483648L-1L)`. Leading `-1` could be ignored when transformed to `BigInt` from `Int` or `Long`.

Reply:
[discussion]To the second one, -2147483648L-1L is valid in long but overflow in int, so we accept this case to non-overflow calculation like `-2147483648-1 = 2147483647`
To the third one, if remove leading `-1`, we will get the wrong value. u just suggest to keep the leading `-1` in BigInt, but this will cause some conflicts if it needs to compare the data entry "same"(not equal value) between `BigInt(Array[Byte](-1, -1, -1, -1, 127, -1, -1, -1))` and `BigInt(Array[Byte](-1, 127, -1, -1, -1))` and this will cause some problems in real applications.

Reply:
Right, `Amount`, `Int32` and `Timestamp` are all **non-negative.** But just keep it careful. And all transformations are checked by validators, so I think negative value will return `InvalidDataEntry => Overflow => Fail` in tx status.

And convert from BigInt to Long or Int, we just give the users chances to use a big negative integer to convert into a positive one. I am not sure about the potential risk here.

Anyway, in current usage, we all take this in careful usage, the current solution should be fine to our working contracts.